FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/ForwardedAttachmentsDownloaderWorker.kt

```diff
@@ -288,18 +304,52 @@ class ForwardedAttachmentsDownloaderWorker(context: Context, params: WorkerParam
     return@withContext msgState
   }
 
-  private suspend fun downloadFile(
-    msgEntity: MessageEntity, pubKeys: List<String>?,
-    destFile: File, srcInputStream: InputStream
-  ) =
-    withContext(Dispatchers.IO) {
-      if (msgEntity.isEncrypted == true) {
-        requireNotNull(pubKeys)
-        PgpEncryptAndOrSign.encryptAndOrSign(srcInputStream, destFile.outputStream(), pubKeys)
+  private suspend fun downloadFileAndProcess(
+    destFile: File,
+    srcInputStream: InputStream,
+    shouldBeEncrypted: Boolean,
+    shouldBeDecryptedBeforeProcess: Boolean,
+    publicKeys: List<String>?,
+    secretKeys: PGPSecretKeyRingCollection? = null,
+    protector: SecretKeyRingProtector? = null
+  ) = withContext(Dispatchers.IO) {
+    var tempFileForDecryptionPurposes: File? = null
+    try {
+      val finalSrcInputStream = if (shouldBeDecryptedBeforeProcess) {
+        tempFileForDecryptionPurposes = File.createTempFile("tmp", null, destFile.parentFile)
```

Comment:
If that's an actual file, and it contains plain - not encrypted data, then this is forbidden. We cannot save that locally, even temporarily, even if it's in app's own storage.

Reply:
Are we talking about `tempFileForDecryptionPurposes`? Do you mean I have not to use a temp file for decryption purposes? Have I to do it on the fly?

Reply:
The plain file can never be saved to the device when forwarding.

That means no intermediary files, unless that intermediary file is properly encrypted.

Reply:
Case 1: original message was encrypted, forwarded message will be encrypted too

I can do the following: I have input stream of encrypted file. I decrypt it to byte array (RAM. Actually it is not good for performance), then I encrypt it and save to a temporary file.

Looks good?

Reply:
You can do that. How will you encrypt it? Already with the final public keys? Do you know them at that point?

Reply:
Case 2: original message was encrypted, forwarded message will be NOT encrypted.

As I understand you don't want I save decrypted file as a temp file, am I right?

Reply:
> You can do that. How will you encrypt it? Already with the final public keys? Do you know them at that point?

Yup, I at that point I know recipients and can fetch public keys from the local database

Reply:
yes, saving the already encrypted file as a temp file is ok

Reply:
Is it necessary to save as a byte array though? Why not stream the whole way through? download -> decrypt -> encrypt -> local file, without keeping all the data in ram.

Reply:
> Why not stream the whole way through? download -> decrypt -> encrypt -> local file

Not sure that it's possible. But before answer I need to check that.

Reply:
Anything is possible :-) the question is more of a level of effort required to do that. I'm also ok to just load it as byte array if that makes things a lot simpler. But we should look into it sometime later to improve it.

Reply:
Case 2. If saving decrypted file is forbidden I will just download an encrypted file at this stage. Then I will decrypt that file when I will send a standard message

Reply:
> Case 2. If saving decrypted file is forbidden I will just download an encrypted file at this stage. Then I will decrypt that file when I will send a standard message

For simplification, it would make sense in both case 1 and 2 to actually save the original encrypted file. Then in one case, it gets only decrypted, and in one, decrypted+re-encrypted. Makes the initial logic same for both. What do you think?

Reply:
> Anything is possible

I wanted to say that it's possible for `Java/Kotlin` world. Just not sure that it's possible for `PGPainless/Bouncy Castle`. Anyway, I will check it tomorrow.

Reply:
> 

Hha, you are totally right 👍. It's a good idea. I can just move all this logic to the next step. As for me manipulation with attachments will be better to do at this stage, to be much clearer. Then the next step will be responsible just for sending... But security reasons encourage me to do file manipulation at the last stage(sending).

Reply:
When I talked about steps I meant the following: we have 3 different services

1. Service that creates MIME message
2. Service that downloads forwarded atts(all things in the current PR relate to it)
3. Service that sends MIME messages.

Reply:
I also think it' good idea. Then you don't need to re-encrypt encrypted message. If you were about to re-encrypt, you would have to generate random temporary key for symmetric encryption algorithm (for example, AES-256), held only in memory, and then encrypt data with it, then later decrypt with it. Not sure if BC is required for this, JDK itself should be capable of symmetric encryption/decryption with some widely accepted algorithms.

Reply:
In the context of forwarding encrypted attachments, the re-encrypt here means encrypting it with the recipient's public keys. Since that's what you have to do before sending it out.